src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/PostgresDistributedLockDAO.scala

```diff
@@ -0,0 +1,128 @@
+package org.broadinstitute.dsde.workbench.sam.dataAccess
+
+import cats.data.OptionT
+import cats.effect.{Async, Resource}
+import cats.syntax.all._
+import fs2.Stream
+import org.broadinstitute.dsde.workbench.model._
+import org.broadinstitute.dsde.workbench.sam.db.SamParameterBinderFactory._
+import org.broadinstitute.dsde.workbench.sam.db._
+import org.broadinstitute.dsde.workbench.sam.db.tables._
+import org.broadinstitute.dsde.workbench.sam.util.lock.DistributedLockAlgebra
+import org.broadinstitute.dsde.workbench.sam.util.DatabaseSupport
+
+import java.time.Instant
+import java.time.temporal.ChronoUnit
+import scala.concurrent.duration.FiniteDuration
+
+class PostgresDistributedLockDAO[F[_]](
+  protected val writeDbRef: DbReference,
+  protected val readDbRef: DbReference,
+  config: DistributedLockConfig
+)(implicit F: Async[F]) extends DistributedLockAlgebra[F] with DatabaseSupport {
+
+  val retryable: Throwable => Boolean = {
+    case _: FailToObtainLock => true
+    case _                   => false
+  }
+
+  def withLock(lock: LockDetails): Resource[F, Unit] = {
+      Resource.make {
+        Stream
+          .retry[F, Unit](acquireLock(lock), config.retryInterval, identity, config.maxRetry, retryable)
+          .adaptError { case e => new WorkbenchException(s"Reached max retry: $e") }
+          .compile
+          .drain
+      }(_ => releaseLock(lock))
+  }
+
+  private[dsde] def acquireLock(lock: LockDetails): F[Unit] = {
+    for {
+      lockStatus <- getLockStatus(lock)
+      _ <- lockStatus match {
+        case Available =>
+          setLock(lock).recoverWith(_ => F.raiseError[Unit](FailToObtainLock(lock)))
+        case Locked    => F.raiseError[Unit](FailToObtainLock(lock))
+      }
+    } yield ()
+  }
+
+  private[dsde] def getLockStatus(lock: LockDetails): F[LockStatus] = {
+    val res = for {
+      lockRecord <- OptionT(F.delay(retrieveLock(lock)))
+      expiresAt <- OptionT.fromOption[F](Option(lockRecord.expiresAt))
+      statusF = F.realTimeInstant.map[LockStatus] { currentTime =>
+        if (currentTime.isAfter(expiresAt)) {
+          deleteLock(lock)
```

Comment:
I think there is a race condition here: 2 threads detect an expired lock, the first proceeds to delete it then re-establish it, then the second thread deletes it.

I think reading and writing state with the same serializable transactions to would help

Reply:
Good point. Moving the transaction up one layer will fix this.

Reply:
For some reason, moving the transaction up a layer caused weird connection issues in the tests. Thank goodness for tests! I've moved the lock-expiry-check logic into the actual DAO method, so that the get/delete can happen within the same transaction, and `retrieveLock` will only return an unexpired lock, or nothing.